[vcpkg] Turn on tests in CI.

BillyONeal opened the pull request.

BillyONeal committed:
[vcpkg] Fix bogus null termination check in unicode decoding.

BillyONeal committed:
[vcpkg] Turn on tests for Linux and OSX.

BillyONeal committed:
[vcpkg] Add Windows tests to CI.

BillyONeal commented:
Example failure: https://dev.azure.com/vcpkg/public/_build/results?buildId=36291&view=logs&j=657c9d73-2672-59f2-b256-fdad23962c37&t=db425a18-2931-5c88-9ee2-7d283bf797fd

BillyONeal committed:
[vcpkg] Back off symlink message to stdout instead of stderr to avoid failing CI over this.

BillyONeal commented:
I filed #11240 to track symlink tests not being able to run in our CI system.

strega-nil commented:
I would prefer it if we just ignore the bootstrap scripts for this purpose; there's no reason to not just build with CMake, IMO.

strega-nil commented on scripts/azure-pipelines/osx/azure-pipelines.yml:
```diff
@@ -30,7 +30,9 @@ jobs:
     inputs:
       filePath: bootstrap-vcpkg.sh
-      arguments: '-allowAppleClang'
+      arguments: '-allowAppleClang -buildTests'
```
I know you didn't add this, but `-allowAppleClang` shouldn't be required.

strega-nil commented on scripts/azure-pipelines/windows/run-tests.yml:
```diff
@@ -0,0 +1,25 @@
+        set PATH=downloads\tools\cmake-3.14.0-windows\cmake-3.14.0-win32-x86\bin;downloads\tools\ninja-1.10.0-windows;%PATH%
+        call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\Common7\Tools\VsDevCmd.bat" -arch=x86 -host_arch=x86
+        rmdir /s /q build.x86.debug > nul 2> nul
+        cmake.exe -G Ninja -DCMAKE_BUILD_TYPE=Debug -DCMAKE_BUILD_TYPE=Debug -DBUILD_TESTING=ON -DVCPKG_DEVELOPMENT_WARNINGS=ON -DVCPKG_WARNINGS_AS_ERRORS=ON -DVCPKG_BUILD_FUZZING=ON -B build.x86.debug -S toolsrc
```
```suggestion
        cmake.exe -G Ninja -DCMAKE_BUILD_TYPE=Debug -DBUILD_TESTING=ON -DVCPKG_DEVELOPMENT_WARNINGS=ON -DVCPKG_WARNINGS_AS_ERRORS=ON -DVCPKG_BUILD_FUZZING=ON -B build.x86.debug -S toolsrc
```

Also, why are you building fuzzing? just to make sure it builds?

BillyONeal replied:
Yes.

BillyONeal committed:
Deduplicate CMAKE_BUILD_TYPE from Nicole.

Co-authored-by: nicole mazzuca <mazzucan@outlook.com>

BillyONeal commented:
> I would prefer it if we just ignore the bootstrap scripts for this purpose; there's no reason to not just build with CMake, IMO.

This is a policy question vis a vis version of cmake we work with. The version of cmake that is on the system is much older than the one bootstrap will install. Do we want to work with the version of cmake that comes with ubuntu 18.04?

strega-nil commented:
Hmm, fair point. We could use pip to install an up to date version tho

strega-nil commented:
/azp run

BillyONeal committed:
Fix static analysis warnings.

CMakeLists.txt: Turn on /analyze when development warnings are on.
catch.hpp: See submitted upstream: https://github.com/catchorg/Catch2/pull/1925
downloads.cpp: PREfast can't see that check_exit is noreturn when f == nullptr.
files.cpp: PREfast detected this bogus both sides of the && testing the same thing.
strings.cpp: Same as the catch issue, isspace needs to be given unsigned.
system.process.cpp: suppress ownership transfer warning about process_info, and make ProcessInfo ensure handles are closed.
build.cpp: PREfast still doesn't understand check_exit
commands.integrate.cpp: More bad <ctype.h> calls.

BillyONeal committed:
Workaround missing _Analysis_assume_ on Linux and MacOS.

BillyONeal commented:
@ras0219 @ras0219-msft Assigning this to you because those test failures are preexisting and I think you're the only one from around when that code was written.

strega-nil commented on toolsrc/src/vcpkg/base/system.process.cpp:
```diff
@@ -365,16 +383,25 @@ namespace vcpkg
+
+#if defined(_MSC_VER)
+#pragma warning(suppress : 6335) // Leaking process information handle 'process_info.proc_info.hProcess'
+                                 // /analyze can't tell that we transferred ownership here
```
Would it be possible to do something like:

```c++
MSVC_SUPPRESS_WARNING(6335)
```

as opposed to this `#if/#endif` thing we do now?

BillyONeal replied:
That would require the ability to emit the pragma in a macro and I'm unaware of reliable ways to do that.

strega-nil replied:
`_Pragma`?

strega-nil commented on toolsrc/src/vcpkg/base/system.process.cpp:
```diff
@@ -327,29 +327,47 @@ namespace vcpkg
+            other.proc_info.hProcess = nullptr;
+            other.proc_info.hThread = nullptr;
+        }
+        ~ProcessInfo()
```
Thank you so much for this

strega-nil commented on toolsrc/src/vcpkg/metrics.cpp:
```diff
@@ -248,6 +248,9 @@ namespace vcpkg::Metrics
+#if defined(_MSC_VER)
+#pragma warning(suppress : 6326) // Potential comparison of a constant with another constant.
+#endif // _MSC_VER
     bool get_compiled_metrics_enabled() { return VCPKG_DISABLE_METRICS == 0; }
```
Maybe this would be better as `return !VCPKG_DISABLE_METRICS;`?

BillyONeal replied:
Good idea!

strega-nil reviewed (approved):
Minor suggestions, but this looks good :)

BillyONeal commented on toolsrc/src/vcpkg/base/system.process.cpp:
```diff
@@ -327,29 +327,47 @@ namespace vcpkg
+
+        friend void swap(ProcessInfo& lhs, ProcessInfo& rhs) noexcept { lhs.swap(rhs); }
 
+        unsigned int wait_and_close_handles()
```
This should just be called wait now

strega-nil committed:
minor fixes

strega-nil commented:
@BillyONeal PR for change ideas at BillyONeal#1

BillyONeal commented:
@strega-nil Merged; also feel free to just push into it if you want :)

BillyONeal committed:
De-allowAppleClang as suggested by Nicole.

BillyONeal commented:
Interestingly one of the failing tests says "// TODO: order here may change but A < X, and B anywhere" -- maybe we are seeing differences in std::sort between the standard libraries here?

BillyONeal committed:
Add sal.h to restore _Analysis_assume_ correctness.

BillyONeal committed:
Fix -Werror=reorder on Linux.

BillyONeal committed:
Delete unused sort bits from json.h.

BillyONeal committed:
Fix set_environment_variable which was still broken 2 tries in.

BillyONeal committed:
Use map rather than unordered_map in places to give consistent topological package sorting across standard library implementations.

BillyONeal commented:
Map fixed it (but Linux is on the floor atm)

BillyONeal committed:
Fix name of wait_and_close_handles()

BillyONeal commented:
/azp run
